### ISSUE TITLE
3.2.4 RC4 modulate and tilemap inconsistencies

### ISSUE THREAD

**Issue by chepulis** [godotengine/godot#46873]
Godot 3.2.4 RC4. Also observed in RC3.

MacOs Big Sur 11.2.2 (20D80), Mid 2015 Retina Macbook 15'. AMD Radeon R9 M370X.
Both GLES3 and GLES2 display issues. Has also showed up on Android (GLES3).

**Issue description:**
Watch the [5 minute video for a demonstration](https://www.youtube.com/watch?v=zaSTcj_Lz7Y).

A 2D shader that consists of this:
```
COLOR = vec4(1.0);
COLOR *= MODULATE;
```

Displays odd behavior, especially on tilemaps:
* Two colors show up in the tilemap, depending on how filled-in it is.
* Sometimes the color might change based on the zoom level in editor.
* It seems that in GLES2 the darker one is the normal modulate, the light is no modulate. In GLES3 the dark is... double modulate? No idea. The light color is the regular modulate. I think.

The change seems to depend around the center of the object (though it does happens further out as well). Moved object retains the issue. Non-tilemap sprites do share at least a part of the issue.

GLES3:
<img width="988" alt="image" src="https://user-images.githubusercontent.com/11724303/110659959-bd58f500-81cb-11eb-8901-6f69c21b33fe.png">

GLES3: Colors are darker, as if modulate happened twice:
<img width="637" alt="image" src="https://user-images.githubusercontent.com/11724303/110659540-59cec780-81cb-11eb-90c9-e0b112c85d48.png">

GLES2: Colors seem to be accurate, but the issue with only part of the tilemap being modulated persists:
<img width="663" alt="image" src="https://user-images.githubusercontent.com/11724303/110659566-5fc4a880-81cb-11eb-8002-a9c7b0aed737.png">

**Steps to reproduce and a testing project**
Have a simple tilemap. Give it the shader material with the shader as above. Give it a *modulate* value. Draw around, watch it break, un-break, re-break.

I've made a [video where i reproduced the issue](https://www.youtube.com/watch?v=zaSTcj_Lz7Y). The project from that video: [tilemap_shader.zip](https://github.com/godotengine/godot/files/6117230/tilemap_shader.zip)

**Comment by lawnjelly** [comment]
I've finally tracked this down. It was quite obscure, and is related to one of the new batching abilities for custom shaders in 3.2.4 : Modulate can be baked into the vertex format using the USE_ATTRIB_MODULATE define in the shader.

The only problem is that the custom shader was incorrectly reading from `final_modulate`, which is the per batch modulate, rather than the attribute modulate. I've got it working now with some bodge code but will clean it up for a PR.

The existing godot logic for MODULATE is quite complex (if it is used in the shader, it relies on you applying it, and ignores the automatic application later in the shader). It's been the source of several bugs, but on the flip side it does give a fair amount of user flexibility.
